## <u>2. Datenaufbereitung Previous Application


*Vorgehensweise kategorische Variablen:*
- Gruppierung der Variablen
- Löschung von kategorischen Variablen mit weniger als 5pP relativer Anteilsdifferenz zwischen Paybacks und Defaults
- Löschung von Variablen mit mehr als 60% fehlenden Daten

## Initialisierung

In [1]:
from pathlib import Path
from scipy import stats

import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns

np.set_printoptions(suppress=True)

pd.options.display.max_columns = None
pd.options.display.max_rows = None
pd.options.display.max_colwidth = None

from sklearn.linear_model import LogisticRegression

from IPython.display import display, Markdown

In [2]:
path1 = Path(r"A:\Workspace\Python\Masterarbeit\Kaggle Home Credit Datensatz")
path2 = Path(r"C:\Users\rober\Documents\Workspace\Python\Masterarbeit\Kaggle Home Credit Datensatz")

if path1.is_dir():
    DATASET_DIR = path1
else:
    DATASET_DIR = path2

In [3]:
pa = pd.read_csv(DATASET_DIR / "previous_application.csv")
pos = pd.read_csv(DATASET_DIR / "2. Datenaufbereitung" / "pos.csv")
ip = pd.read_csv(DATASET_DIR / "2. Datenaufbereitung" / "ip.csv")
cc = pd.read_csv(DATASET_DIR / "2. Datenaufbereitung" / "cc.csv")
description = pd.read_csv(DATASET_DIR / "HomeCredit_columns_description.csv", encoding="latin", index_col=0)

In [4]:
pa = pa.set_index("SK_ID_PREV")
pos = pos.set_index("SK_ID_PREV")
ip = ip.set_index("SK_ID_PREV")
cc = cc.set_index("SK_ID_PREV")

In [5]:
pa.head()

SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
SK_ID_PREV                                                                
2030495         271877     Consumer loans     1730.430          17145.0   
2802425         108129         Cash loans    25188.615         607500.0   
2523466         122040         Cash loans    15060.735         112500.0   
2819243         176158         Cash loans    47041.335         450000.0   
1784265         202054         Cash loans    31924.395         337500.0   

            AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  \
SK_ID_PREV                                                  
2030495        17145.0               0.0          17145.0   
2802425       679671.0               NaN         607500.0   
2523466       136444.5               NaN         112500.0   
2819243       470790.0               NaN         450000.0   
1784265       404055.0               NaN         337500.0   

           WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  \
SK_ID_PREV                                                       
2030495                      SATURDAY                       15   
2802425                      THURSDAY                       11   
2523466                       TUESDAY                       11   
2819243                        MONDAY                        7   
1784265                      THURSDAY                        9   

           FLAG_LAST_APPL_PER_CONTRACT  NFLAG_LAST_APPL_IN_DAY  \
SK_ID_PREV                                                       
2030495                              Y                       1   
2802425                              Y                       1   
2523466                              Y                       1   
2819243                              Y                       1   
1784265                              Y                       1   

            RATE_DOWN_PAYMENT  RATE_INTEREST_PRIMARY  \
SK_ID_PREV                                             
2030495                   0.0               0.182832   
2802425                   NaN                    NaN   
2523466                   NaN                    NaN   
2819243                   NaN                    NaN   
1784265                   NaN                    NaN   

            RATE_INTEREST_PRIVILEGED NAME_CASH_LOAN_PURPOSE  \
SK_ID_PREV                                                    
2030495                     0.867336                    XAP   
2802425                          NaN                    XNA   
2523466                          NaN                    XNA   
2819243                          NaN                    XNA   
1784265                          NaN                Repairs   

           NAME_CONTRACT_STATUS  DAYS_DECISION      NAME_PAYMENT_TYPE  \
SK_ID_PREV                                                              
2030495                Approved            -73  Cash through the bank   
2802425                Approved           -164                    XNA   
2523466                Approved           -301  Cash through the bank   
2819243                Approved           -512  Cash through the bank   
1784265                 Refused           -781  Cash through the bank   

           CODE_REJECT_REASON  NAME_TYPE_SUITE NAME_CLIENT_TYPE  \
SK_ID_PREV                                                        
2030495                   XAP              NaN         Repeater   
2802425                   XAP    Unaccompanied         Repeater   
2523466                   XAP  Spouse, partner         Repeater   
2819243                   XAP              NaN         Repeater   
1784265                    HC              NaN         Repeater   

           NAME_GOODS_CATEGORY NAME_PORTFOLIO NAME_PRODUCT_TYPE  \
SK_ID_PREV                                                        
2030495                 Mobile            POS               XNA   
2802425                    XNA           Cash            x-sell   
2523466                    XNA           Cash  

In [6]:
# Spalten die innerhalb der Aufbereitung nicht verändert werden können
skip = ["SK_ID_CURR", "SK_ID_PREV"]

In [7]:
# nominale und metrische Spalten
n_heads = [element for element in pa.columns if pa[element].dtype.name == "object"]
n_heads += ["NFLAG_LAST_APPL_IN_DAY", "NFLAG_INSURED_ON_APPROVAL"]
o_heads = ["HOUR_APPR_PROCESS_START", "SELLERPLACE_AREA"]
m_heads = [element for element in pa.columns if element not in n_heads and element not in o_heads]

## <u> kategorische Variablen

In [8]:
df = pa[["SK_ID_CURR"] + n_heads].copy()

In [9]:
result = pd.DataFrame(index=pa.SK_ID_CURR.unique())
result.index.name = "SK_ID_CURR"

In [10]:
df.head()

SK_ID_CURR NAME_CONTRACT_TYPE WEEKDAY_APPR_PROCESS_START  \
SK_ID_PREV                                                             
2030495         271877     Consumer loans                   SATURDAY   
2802425         108129         Cash loans                   THURSDAY   
2523466         122040         Cash loans                    TUESDAY   
2819243         176158         Cash loans                     MONDAY   
1784265         202054         Cash loans                   THURSDAY   

           FLAG_LAST_APPL_PER_CONTRACT NAME_CASH_LOAN_PURPOSE  \
SK_ID_PREV                                                      
2030495                              Y                    XAP   
2802425                              Y                    XNA   
2523466                              Y                    XNA   
2819243                              Y                    XNA   
1784265                              Y                Repairs   

           NAME_CONTRACT_STATUS      NAME_PAYMENT_TYPE CODE_REJECT_REASON  \
SK_ID_PREV                                                                  
2030495                Approved  Cash through the bank                XAP   
2802425                Approved                    XNA                XAP   
2523466                Approved  Cash through the bank                XAP   
2819243                Approved  Cash through the bank                XAP   
1784265                 Refused  Cash through the bank                 HC   

            NAME_TYPE_SUITE NAME_CLIENT_TYPE NAME_GOODS_CATEGORY  \
SK_ID_PREV                                                         
2030495                 NaN         Repeater              Mobile   
2802425       Unaccompanied         Repeater                 XNA   
2523466     Spouse, partner         Repeater                 XNA   
2819243                 NaN         Repeater                 XNA   
1784265                 NaN         Repeater                 XNA   

           NAME_PORTFOLIO NAME_PRODUCT_TYPE             CHANNEL_TYPE  \
SK_ID_PREV                                                             
2030495               POS               XNA             Country-wide   
2802425              Cash            x-sell           Contact center   
2523466              Cash            x-sell  Credit and cash offices   
2819243              Cash            x-sell  Credit and cash offices   
1784265              Cash           walk-in  Credit and cash offices   

           NAME_SELLER_INDUSTRY NAME_YIELD_GROUP       PRODUCT_COMBINATION  \
SK_ID_PREV                                                                   
2030495            Connectivity           middle  POS mobile with interest   
2802425                     XNA       low_action          Cash X-Sell: low   
2523466                     XNA             high         Cash X-Sell: high   
2819243                     XNA           middle       Cash X-Sell: middle   
1784265                     XNA             high         Cash Street: high   

            NFLAG_LAST_APPL_IN_DAY  NFLAG_INSURED_ON_APPROVAL  
SK_ID_PREV                                                     
2030495                          1                        0.0  
2802425                          1                        1.0  
2523466                          1                        1.0  
2819243                          1                        1.0  
1784265                          1                        NaN

### Aufsummierung aller historischen Kredite nach Kategorie

In [11]:
for cat in df.columns.values:
    if cat in skip:
        continue
    
    A = df[["SK_ID_CURR", cat]].groupby(by=["SK_ID_CURR", cat]).size().unstack(fill_value=0)
    A.columns = ["CNT_{cat}_{head}".format(cat=cat, head=head) for head in A.keys().values]
    result = pd.merge(result, A, how="left", left_index=True, right_index=True)

In [12]:
df = result
df.head()

CNT_NAME_CONTRACT_TYPE_Cash loans  \
SK_ID_CURR                                      
271877                                      0   
108129                                      2   
122040                                      2   
176158                                     22   
202054                                     18   

            CNT_NAME_CONTRACT_TYPE_Consumer loans  \
SK_ID_CURR                                          
271877                                          3   
108129                                          3   
122040                                          1   
176158                                          1   
202054                                          6   

            CNT_NAME_CONTRACT_TYPE_Revolving loans  \
SK_ID_CURR                                           
271877                                           0   
108129                                           1   
122040                                           1   
176158                                           0   
202054                                           1   

            CNT_NAME_CONTRACT_TYPE_XNA  CNT_WEEKDAY_APPR_PROCESS_START_FRIDAY  \
SK_ID_CURR                                                                      
271877                               0                                      0   
108129                               0                                      0   
122040                               0                                      1   
176158                               0                                      3   
202054                               0                                      4   

            CNT_WEEKDAY_APPR_PROCESS_START_MONDAY  \
SK_ID_CURR                                          
271877                                          0   
108129                                          0   
122040                                          0   
176158                                         10   
202054                                          6   

            CNT_WEEKDAY_APPR_PROCESS_START_SATURDAY  \
SK_ID_CURR                                            
271877                                            2   
108129                                            0   
122040                                            0   
176158                                            1   
202054                                            2   

            CNT_WEEKDAY_APPR_PROCESS_START_SUNDAY  \
SK_ID_CURR                                          
271877                                          1   
108129                                          1   
122040                                          0   
176158                                          0   
202054                                          1   

            CNT_WEEKDAY_APPR_PROCESS_START_THURSDAY  \
SK_ID_CURR                                            
271877                                            0   
108129                                            2   
122040                                            0   
176158                                            1   
202054                                            8   

            CNT_WEEKDAY_APPR_PROCESS_START_TUESDAY  \
SK_ID_CURR                                           
271877                                           0   
108129                                           1   
122040                                           3   
176158                                           7   
202054                                           4   

            CNT_WEEKDAY_APPR_PROCESS_START_WEDNESDAY  \
SK_ID_CURR                                             
271877                                             0   
108129                                             2   
122040                                             0   
176158                                             1   
202054                                             0   

            CNT_FLAG_LAST_APPL_PER_CONTRACT_N  \
SK_ID_CURR  

Kreditnehmer 271877 hat in seiner Kredithistorie von Home Credit 3 Konsumentenkredite beantragt, wovon 2 genehmigt und 1 abgelehnt wurde, etc. 

### Speichern der kategorischen Werte

In [13]:
cats = df

## <u> metrische Variablen

In [14]:
ID_PA_POS = np.intersect1d(pa.index.values, pos.index.values)
ID_PA_IP = np.intersect1d(pa.index.values, ip.index.values)
ID_PA_CC = np.intersect1d(pa.index.values, cc.index.values)

# Ermittlung der Kredite ohne Kredithistorie
# -> PA's, die nicht in POS, IP oder CC sind

A = np.concatenate((ID_PA_POS,ID_PA_IP,ID_PA_CC)) 
B = pa.drop(A)

ID_PA_ONLY = B.index.values

In [15]:
# Erstellung der Unterdatensätze je nach Kredittyp

In [16]:
pa_pos = pa.loc[ID_PA_POS]
pa_ip = pa.loc[ID_PA_IP]
pa_cc = pa.loc[ID_PA_CC]
pa_only = pa.loc[ID_PA_ONLY]

In [17]:
pa_pos = pd.merge(pa_pos, pos, how="left", left_index=True, right_index=True)
pa_ip = pd.merge(pa_ip, ip, how="left", left_index=True, right_index=True)
pa_cc = pd.merge(pa_cc, cc, how="left", left_index=True, right_index=True)

In [18]:
df.head()

CNT_NAME_CONTRACT_TYPE_Cash loans  \
SK_ID_CURR                                      
271877                                      0   
108129                                      2   
122040                                      2   
176158                                     22   
202054                                     18   

            CNT_NAME_CONTRACT_TYPE_Consumer loans  \
SK_ID_CURR                                          
271877                                          3   
108129                                          3   
122040                                          1   
176158                                          1   
202054                                          6   

            CNT_NAME_CONTRACT_TYPE_Revolving loans  \
SK_ID_CURR                                           
271877                                           0   
108129                                           1   
122040                                           1   
176158                                           0   
202054                                           1   

            CNT_NAME_CONTRACT_TYPE_XNA  CNT_WEEKDAY_APPR_PROCESS_START_FRIDAY  \
SK_ID_CURR                                                                      
271877                               0                                      0   
108129                               0                                      0   
122040                               0                                      1   
176158                               0                                      3   
202054                               0                                      4   

            CNT_WEEKDAY_APPR_PROCESS_START_MONDAY  \
SK_ID_CURR                                          
271877                                          0   
108129                                          0   
122040                                          0   
176158                                         10   
202054                                          6   

            CNT_WEEKDAY_APPR_PROCESS_START_SATURDAY  \
SK_ID_CURR                                            
271877                                            2   
108129                                            0   
122040                                            0   
176158                                            1   
202054                                            2   

            CNT_WEEKDAY_APPR_PROCESS_START_SUNDAY  \
SK_ID_CURR                                          
271877                                          1   
108129                                          1   
122040                                          0   
176158                                          0   
202054                                          1   

            CNT_WEEKDAY_APPR_PROCESS_START_THURSDAY  \
SK_ID_CURR                                            
271877                                            0   
108129                                            2   
122040                                            0   
176158                                            1   
202054                                            8   

            CNT_WEEKDAY_APPR_PROCESS_START_TUESDAY  \
SK_ID_CURR                                           
271877                                           0   
108129                                           1   
122040                                           3   
176158                                           7   
202054                                           4   

            CNT_WEEKDAY_APPR_PROCESS_START_WEDNESDAY  \
SK_ID_CURR                                             
271877                                             0   
108129                                             2   
122040                                             0   
176158                                             1   
202054                                             0   

            CNT_FLAG_LAST_APPL_PER_CONTRACT_N  \
SK_ID_CURR  

## PA_POS

In [19]:
# pa_pos

In [20]:
df = pa_pos
df = df.drop(n_heads + o_heads, axis=1)
df.head()

SK_ID_CURR  AMT_ANNUITY  AMT_APPLICATION  AMT_CREDIT  \
SK_ID_PREV                                                         
1000001         158271     6404.310        58905.000     65124.0   
1000002         101962     6264.000        39145.500     35230.5   
1000003         252457     4951.350        47056.275     52641.0   
1000004         260094     3391.110        35144.370     30586.5   
1000005         176456    14713.605       123486.075    120307.5   

            AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  RATE_DOWN_PAYMENT  \
SK_ID_PREV                                                         
1000001                0.000        58905.000           0.000000   
1000002             3915.000        39145.500           0.108922   
1000003                4.275        47056.275           0.000088   
1000004             7032.870        35144.370           0.203603   
1000005            12349.575       123486.075           0.101388   

            RATE_INTEREST_PRIMARY  RATE_INTEREST_PRIVILEGED  DAYS_DECISION  \
SK_ID_PREV                                                                   
1000001                       NaN                       NaN           -299   
1000002                       NaN                       NaN          -1634   
1000003                       NaN                       NaN           -124   
1000004                       NaN                       NaN           -893   
1000005                       NaN                       NaN          -1719   

            CNT_PAYMENT  DAYS_FIRST_DRAWING  DAYS_FIRST_DUE  \
SK_ID_PREV                                                    
1000001            12.0            365243.0          -268.0   
1000002             6.0            365243.0         -1600.0   
1000003            12.0            365243.0           -94.0   
1000004            10.0            365243.0          -862.0   
1000005            10.0            365243.0         -1688.0   

            DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE  DAYS_TERMINATION  \
SK_ID_PREV                                                               
1000001                          62.0         -238.0            -233.0   
1000002                       -1450.0        -1510.0           -1501.0   
1000003                         236.0       365243.0          365243.0   
1000004                        -592.0         -682.0            -672.0   
1000005                       -1418.0        -1418.0           -1415.0   

            POS_CNT_PAYMENTS_LEFT  POS_CNT_DPD  
SK_ID_PREV                                      
1000001                       0.0          0.0  
1000002                       0.0          0.0  
1000003                       9.0          0.0  
1000004                       0.0          0.0  
1000005                       0.0          0.0

In [21]:
result = pd.DataFrame(index=pa_pos.SK_ID_CURR.unique())
result.index.name = "SK_ID_CURR"

### Gruppierung

In [22]:
# count
A = df[["SK_ID_CURR"]]
A.reset_index(level=0, inplace=True)
A = A.groupby(by=["SK_ID_CURR"]).count()
A.columns = ["CNT_CREDITS"]
result = pd.merge(result, A, how="left", left_index=True, right_index=True)

Anzahl der vergangenen POS Kredite

In [23]:
# sums
sum_heads = ["POS_CNT_PAYMENTS_LEFT", "POS_CNT_DPD"] + ["AMT_ANNUITY", "AMT_APPLICATION", "AMT_CREDIT", "AMT_DOWN_PAYMENT", "AMT_GOODS_PRICE", "RATE_DOWN_PAYMENT"]

for head in sum_heads:
    A = df[["SK_ID_CURR", head]].groupby(by=["SK_ID_CURR"]).sum()
    result = pd.merge(result, A, how="left", left_index=True, right_index=True)

Hat ein Kreditnehmer mehrere Kredite paralell werden deren Daten aufsummiert um die gesamte jährliche Zahlungshöhe, Kredithöhe etc zu ermitteln.

### Zusammenfügen und Speichern der PA_POS Daten

In [24]:
pa_pos = result

In [25]:
pa_pos = pd.merge(pa_pos, cats, how="left", left_index=True, right_index=True)

In [26]:
pa_pos = pa_pos.add_prefix("PA_")

In [27]:
pa_pos = pa_pos.sort_index()

In [28]:
pa_pos2 = pd.read_csv(DATASET_DIR / "2. Datenaufbereitung" / "pa_pos.csv",index_col=0)

In [29]:
pa_pos = pa_pos[pa_pos2.columns.values]

In [30]:
pa_pos.to_csv(DATASET_DIR / "2. Datenaufbereitung" / "pa_pos_all.csv")

## PA_IP

In [31]:
# pa_ip

In [32]:
df = pa_ip
df = df.drop(n_heads + o_heads, axis=1)
df.head()

SK_ID_CURR  AMT_ANNUITY  AMT_APPLICATION  AMT_CREDIT  \
SK_ID_PREV                                                         
1000003         252457     4951.350        47056.275     52641.0   
1000007         256657    11246.805        78570.000     62856.0   
1000011         198678    92435.040       855000.000    879831.0   
1000017         310743     7002.720        67864.230     75024.0   
1000018         394447     2250.000        45000.000     45000.0   

            AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  RATE_DOWN_PAYMENT  \
SK_ID_PREV                                                         
1000003                4.275        47056.275           0.000088   
1000007            15714.000        78570.000           0.217818   
1000011                  NaN       855000.000                NaN   
1000017                4.230        67864.230           0.000061   
1000018                  NaN        45000.000                NaN   

            RATE_INTEREST_PRIMARY  RATE_INTEREST_PRIVILEGED  DAYS_DECISION  \
SK_ID_PREV                                                                   
1000003                       NaN                       NaN           -124   
1000007                       NaN                       NaN           -158   
1000011                       NaN                       NaN           -465   
1000017                       NaN                       NaN           -516   
1000018                       NaN                       NaN           -176   

            CNT_PAYMENT  DAYS_FIRST_DRAWING  DAYS_FIRST_DUE  \
SK_ID_PREV                                                    
1000003            12.0            365243.0           -94.0   
1000007             6.0            365243.0          -123.0   
1000011            12.0            365243.0          -435.0   
1000017            12.0            365243.0          -486.0   
1000018             0.0              -175.0          -130.0   

            DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE  DAYS_TERMINATION  \
SK_ID_PREV                                                               
1000003                         236.0       365243.0          365243.0   
1000007                          27.0       365243.0          365243.0   
1000011                        -105.0         -105.0            -101.0   
1000017                        -156.0         -156.0            -151.0   
1000018                      365243.0       365243.0          365243.0   

            IP_PAYMENT_30  IP_DAYS_DIFF_30  IP_AMT_DIFF_30  IP_PAYMENT_60  \
SK_ID_PREV                                                                  
1000003         -0.912871              0.0             0.0       0.912871   
1000007          0.408248             -7.0             0.0       0.408248   
1000011         -0.912871              0.0             0.0      -0.912871   
1000017         -0.408248              0.0             0.0      -0.408248   
1000018          0.650875              0.0             0.0       0.814828   

            IP_DAYS_DIFF_60  IP_AMT_DIFF_60  IP_PAYMENT_90  IP_DAYS_DIFF_90  \
SK_ID_PREV                                                                    
1000003               -15.0             0.0       0.912871            -17.0   
1000007               -21.0             0.0       0.408248            -17.0   
1000011                 0.0             0.0      -0.912871              0.0   
1000017                 0.0             0.0      -0.408248              0.0   
1000018                -1.0             0.0      -1.044958             -1.0   

            IP_AMT_DIFF_90  IP_PAYMENT_120  IP_DAYS_DIFF_120  IP_AMT_DIFF_120  \
SK_ID_PREV                                                                      
1000003                0.0        0.912871             -14.0              0.0   
1000007                0.0        0.408248             -19.0              0.0   
1000011                0.0        0.912056              -2.0              0.0   
1000017                0.0       -0.408248         

In [33]:
result = pd.DataFrame(index=pa_ip.SK_ID_CURR.unique())
result.index.name = "SK_ID_CURR"

In [34]:
# count
A = df[["SK_ID_CURR"]]
A.reset_index(level=0, inplace=True)
A = A.groupby(by=["SK_ID_CURR"]).count()
A.columns = ["CNT_CREDITS"]
result = pd.merge(result, A, how="left", left_index=True, right_index=True)

Anzahl der vergangenen Install Payments Kredite

In [35]:
ip.columns.values

array(['IP_PAYMENT_30', 'IP_DAYS_DIFF_30', 'IP_AMT_DIFF_30',
       'IP_PAYMENT_60', 'IP_DAYS_DIFF_60', 'IP_AMT_DIFF_60',
       'IP_PAYMENT_90', 'IP_DAYS_DIFF_90', 'IP_AMT_DIFF_90',
       'IP_PAYMENT_120', 'IP_DAYS_DIFF_120', 'IP_AMT_DIFF_120',
       'IP_PAYMENT_150', 'IP_DAYS_DIFF_150', 'IP_AMT_DIFF_150',
       'IP_PAYMENT_180', 'IP_DAYS_DIFF_180', 'IP_AMT_DIFF_180'],
      dtype=object)

In [36]:
# means
mean_heads = list(ip.columns.values)

for head in mean_heads:
    A = df[["SK_ID_CURR", head]]
    A = A.fillna(0)
    A = A.groupby(by=["SK_ID_CURR"]).mean()
    result = pd.merge(result, A, how="left", left_index=True, right_index=True)

Aus mehereren Zeitreihen wird der Durchschnitt gebildet, da die Daten normalisiert sind

In [37]:
# sums
sums_heads = ["AMT_ANNUITY", "AMT_APPLICATION", "AMT_CREDIT", "AMT_DOWN_PAYMENT", "AMT_GOODS_PRICE", "RATE_DOWN_PAYMENT"]

for head in sums_heads:
    A = df[["SK_ID_CURR", head]]
    A = A.fillna(0)
    A = A.groupby(by=["SK_ID_CURR"]).sum()
    result = pd.merge(result, A, how="left", left_index=True, right_index=True)

Hat ein Kreditnehmer mehrere Kredite paralell werden deren Daten aufsummiert um die gesamte jährliche Zahlungshöhe, Kredithöhe etc zu ermitteln.

In [38]:
result.head()

CNT_CREDITS  IP_PAYMENT_30  IP_DAYS_DIFF_30  IP_AMT_DIFF_30  \
SK_ID_CURR                                                                
252457                1  -9.128709e-01              0.0             0.0   
256657                2   2.041241e-01             -6.0             0.0   
198678                2   4.540312e-08            -20.0             0.0   
310743                1  -4.082483e-01              0.0             0.0   
394447                2   5.295619e-01             -0.5             0.0   

            IP_PAYMENT_60  IP_DAYS_DIFF_60  IP_AMT_DIFF_60  IP_PAYMENT_90  \
SK_ID_CURR                                                                  
252457       9.128709e-01            -15.0             0.0   9.128709e-01   
256657       2.041241e-01            -13.0             0.0   2.041241e-01   
198678       4.540312e-08            -21.5             0.0   4.540312e-08   
310743      -4.082483e-01              0.0             0.0  -4.082483e-01   
394447       6.115381e-01             -7.0             0.0  -3.183548e-01   

            IP_DAYS_DIFF_90  IP_AMT_DIFF_90  IP_PAYMENT_120  IP_DAYS_DIFF_120  \
SK_ID_CURR                                                                      
252457                -17.0             0.0        0.912871             -14.0   
256657                -12.0             0.0        0.204124             -12.5   
198678                 -6.5             0.0       -0.000407              -1.0   
310743                  0.0             0.0       -0.408248               0.0   
394447                 -7.0             0.0       -0.229629              -6.5   

            IP_AMT_DIFF_120  IP_PAYMENT_150  IP_DAYS_DIFF_150  \
SK_ID_CURR                                                      
252457                  0.0       -0.912871               0.0   
256657                  0.0        0.204124             -13.0   
198678                  0.0        0.000204               0.5   
310743                  0.0       -0.408248               0.0   
394447                  0.0       -0.180184              -8.5   

            IP_AMT_DIFF_150  IP_PAYMENT_180  IP_DAYS_DIFF_180  \
SK_ID_CURR                                                      
252457                  0.0       -0.912871               0.0   
256657                  0.0       -1.020621              -1.5   
198678                  0.0        0.000204              -0.5   
310743                  0.0        2.041241              -5.0   
394447                  0.0       -0.412932               0.0   

            IP_AMT_DIFF_180  AMT_ANNUITY  AMT_APPLICATION  AMT_CREDIT  \
SK_ID_CURR                                                              
252457                  0.0     4951.350        47056.275     52641.0   
256657                  0.0    49054.770       978570.000   1198062.0   
198678                  0.0    98806.050       921721.500    941661.0   
310743                  0.0     7002.720        67864.230     75024.0   
394447                  0.0    17141.445       129132.000    129132.0   

            AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  RATE_DOWN_PAYMENT  
SK_ID_CURR                                                        
252457                 4.275        47056.275           0.000088  
256657             15714.000       978570.000           0.217818  
198678             13347.000       921721.500           0.193358  
310743                 4.230        67864.230           0.000061  
394447                 0.000       129132.000           0.000000

### Zusammenführen und Speichern der Daten

In [39]:
pa_ip = result

In [40]:
pa_ip = pd.merge(pa_ip, cats, how="left", left_index=True, right_index=True)

In [41]:
pa_ip = pa_ip.add_prefix("PA_")

In [42]:
pa_ip2 = pd.read_csv(DATASET_DIR / "2. Datenaufbereitung" / "pa_ip.csv",index_col=0)

In [43]:
pa_ip = pa_ip[pa_ip2.columns.values]

In [44]:
pa_ip.head()

PA_CNT_CREDITS  PA_IP_PAYMENT_30  PA_IP_DAYS_DIFF_30  \
SK_ID_CURR                                                         
252457                   1     -9.128709e-01                 0.0   
256657                   2      2.041241e-01                -6.0   
198678                   2      4.540312e-08               -20.0   
310743                   1     -4.082483e-01                 0.0   
394447                   2      5.295619e-01                -0.5   

            PA_IP_AMT_DIFF_30  PA_IP_PAYMENT_60  PA_IP_DAYS_DIFF_60  \
SK_ID_CURR                                                            
252457                    0.0      9.128709e-01               -15.0   
256657                    0.0      2.041241e-01               -13.0   
198678                    0.0      4.540312e-08               -21.5   
310743                    0.0     -4.082483e-01                 0.0   
394447                    0.0      6.115381e-01                -7.0   

            PA_IP_AMT_DIFF_60  PA_IP_PAYMENT_90  PA_IP_DAYS_DIFF_90  \
SK_ID_CURR                                                            
252457                    0.0      9.128709e-01               -17.0   
256657                    0.0      2.041241e-01               -12.0   
198678                    0.0      4.540312e-08                -6.5   
310743                    0.0     -4.082483e-01                 0.0   
394447                    0.0     -3.183548e-01                -7.0   

            PA_IP_AMT_DIFF_90  PA_IP_PAYMENT_120  PA_IP_DAYS_DIFF_120  \
SK_ID_CURR                                                              
252457                    0.0           0.912871                -14.0   
256657                    0.0           0.204124                -12.5   
198678                    0.0          -0.000407                 -1.0   
310743                    0.0          -0.408248                  0.0   
394447                    0.0          -0.229629                 -6.5   

            PA_IP_AMT_DIFF_120  PA_IP_PAYMENT_150  PA_IP_DAYS_DIFF_150  \
SK_ID_CURR                                                               
252457                     0.0          -0.912871                  0.0   
256657                     0.0           0.204124                -13.0   
198678                     0.0           0.000204                  0.5   
310743                     0.0          -0.408248                  0.0   
394447                     0.0          -0.180184                 -8.5   

            PA_IP_AMT_DIFF_150  PA_IP_PAYMENT_180  PA_IP_DAYS_DIFF_180  \
SK_ID_CURR                                                               
252457                     0.0          -0.912871                  0.0   
256657                     0.0          -1.020621                 -1.5   
198678                     0.0           0.000204                 -0.5   
310743                     0.0           2.041241                 -5.0   
394447                     0.0          -0.412932                  0.0   

            PA_IP_AMT_DIFF_180  PA_AMT_ANNUITY  PA_AMT_APPLICATION  \
SK_ID_CURR                                                           
252457                     0.0        4951.350           47056.275   
256657                     0.0       49054.770          978570.000   
198678                     0.0       98806.050          921721.500   
310743                     0.0        7002.720           67864.230   
394447                     0.0       17141.445          129132.000   

            PA_AMT_CREDIT  PA_AMT_DOWN_PAYMENT  PA_AMT_GOODS_PRICE  \
SK_ID_CURR                                                           
252457            52641.0                4.275           47056.275   
256657          1198062.0            15714.000          978570.000   
198678           941661.0            13347.000          921721.500   
310743            75024.0                4.230           67864.230   
394447           129132.0                0.000          129132.000   

       

In [45]:
pa_ip.to_csv(DATASET_DIR / "2. Datenaufbereitung" / "pa_ip_all.csv")

## PA_CC

In [46]:
# pa_cc

In [47]:
n_heads.remove("NAME_TYPE_SUITE")

In [48]:
df = pa_cc
df = df.drop(n_heads + o_heads, axis=1)
df.head()

SK_ID_CURR  AMT_ANNUITY  AMT_APPLICATION  AMT_CREDIT  \
SK_ID_PREV                                                         
1000018         394447       2250.0          45000.0     45000.0   
1000030         361282       2250.0          45000.0     45000.0   
1000031         131335       2250.0          45000.0     45000.0   
1000035         436351       2250.0          45000.0     45000.0   
1000077         181153       2250.0          45000.0     45000.0   

            AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  RATE_DOWN_PAYMENT  \
SK_ID_PREV                                                         
1000018                  NaN          45000.0                NaN   
1000030                  NaN          45000.0                NaN   
1000031                  NaN          45000.0                NaN   
1000035                  NaN          45000.0                NaN   
1000077                  NaN          45000.0                NaN   

            RATE_INTEREST_PRIMARY  RATE_INTEREST_PRIVILEGED  DAYS_DECISION  \
SK_ID_PREV                                                                   
1000018                       NaN                       NaN           -176   
1000030                       NaN                       NaN           -255   
1000031                       NaN                       NaN           -499   
1000035                       NaN                       NaN           -178   
1000077                       NaN                       NaN           -351   

           NAME_TYPE_SUITE  CNT_PAYMENT  DAYS_FIRST_DRAWING  DAYS_FIRST_DUE  \
SK_ID_PREV                                                                    
1000018      Unaccompanied          0.0              -175.0          -130.0   
1000030             Family          0.0              -243.0          -178.0   
1000031             Family          0.0              -396.0          -362.0   
1000035             Family          0.0            365243.0        365243.0   
1000077                NaN          0.0            365243.0        365243.0   

            DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE  DAYS_TERMINATION  \
SK_ID_PREV                                                               
1000018                      365243.0       365243.0          365243.0   
1000030                      365243.0       365243.0          365243.0   
1000031                      365243.0       365243.0          365243.0   
1000035                      365243.0       365243.0          365243.0   
1000077                      365243.0       365243.0          365243.0   

            CC_BALANCE_30  CC_PAYMENT_30  CC_DPD_30  CC_BALANCE_60  \
SK_ID_PREV                                                           
1000018         -1.206367      -1.389108        0.0       1.434002   
1000030          0.946004       0.850318        0.0       0.828945   
1000031          0.722068      -0.427910        0.0       0.719445   
1000035               NaN            NaN        0.0            NaN   
1000077               NaN            NaN        0.0            NaN   

            CC_PAYMENT_60  CC_DPD_60  CC_BALANCE_90  CC_PAYMENT_90  CC_DPD_90  \
SK_ID_PREV                                                                      
1000018          1.008387        0.0       0.992963      -0.429381        0.0   
1000030          1.109609        0.0       0.861611       0.179508        0.0   
1000031         -0.427508        0.0       0.791575      -0.445538        0.0   
1000035               NaN        0.0            NaN            NaN        0.0   
1000077               NaN        0.0            NaN            NaN        0.0   

            CC_BALANCE_120  CC_PAYMENT_120  CC_DPD_120  CC_BALANCE_150  \
SK_ID_PREV                                                               
1000018          -0.349513       -0.292196         0.0       -0.415697   
1000030          -0.393110       -1.267179         0.0       -1.074370   
1000031           0.204483       -0.215249         0.0       -0.817702   
1000035  

In [49]:
result = pd.DataFrame(index=pa_cc.SK_ID_CURR.unique())
result.index.name = "SK_ID_CURR"

### Gruppierung

In [50]:
# count
A = df[["SK_ID_CURR"]]
A.reset_index(level=0, inplace=True)
A = A.groupby(by=["SK_ID_CURR"]).count()
A.columns = ["CNT_CREDITS"]
result = pd.merge(result, A, how="left", left_index=True, right_index=True)

Anzahl der Kredite im CC-Datensatz je Kreditnehmer

In [51]:
cc.columns.values

array(['CC_BALANCE_30', 'CC_PAYMENT_30', 'CC_DPD_30', 'CC_BALANCE_60',
       'CC_PAYMENT_60', 'CC_DPD_60', 'CC_BALANCE_90', 'CC_PAYMENT_90',
       'CC_DPD_90', 'CC_BALANCE_120', 'CC_PAYMENT_120', 'CC_DPD_120',
       'CC_BALANCE_150', 'CC_PAYMENT_150', 'CC_DPD_150', 'CC_BALANCE_180',
       'CC_PAYMENT_180', 'CC_DPD_180'], dtype=object)

In [52]:
# means
mean_heads = list(cc.columns.values)

for head in mean_heads:
    A = df[["SK_ID_CURR", head]]
    A = A.fillna(0)
    A = A.groupby(by=["SK_ID_CURR"]).mean()
    result = pd.merge(result, A, how="left", left_index=True, right_index=True)

In [53]:
# sums
sum_heads = ["AMT_ANNUITY", "AMT_APPLICATION", "AMT_CREDIT", "AMT_GOODS_PRICE"]

for head in sum_heads:
    A = df[["SK_ID_CURR", head]]
    A = A.fillna(0)
    A = A.groupby(by=["SK_ID_CURR"]).sum()
    result = pd.merge(result, A, how="left", left_index=True, right_index=True)

In [54]:
result.head()

CNT_CREDITS  CC_BALANCE_30  CC_PAYMENT_30  CC_DPD_30  \
SK_ID_CURR                                                         
394447                1      -1.206367      -1.389108        0.0   
361282                1       0.946004       0.850318        0.0   
131335                1       0.722068      -0.427910        0.0   
436351                1       0.000000       0.000000        0.0   
181153                1       0.000000       0.000000        0.0   

            CC_BALANCE_60  CC_PAYMENT_60  CC_DPD_60  CC_BALANCE_90  \
SK_ID_CURR                                                           
394447           1.434002       1.008387        0.0       0.992963   
361282           0.828945       1.109609        0.0       0.861611   
131335           0.719445      -0.427508        0.0       0.791575   
436351           0.000000       0.000000        0.0       0.000000   
181153           0.000000       0.000000        0.0       0.000000   

            CC_PAYMENT_90  CC_DPD_90  CC_BALANCE_120  CC_PAYMENT_120  \
SK_ID_CURR                                                             
394447          -0.429381        0.0       -0.349513       -0.292196   
361282           0.179508        0.0       -0.393110       -1.267179   
131335          -0.445538        0.0        0.204483       -0.215249   
436351           0.000000        0.0        0.000000        0.000000   
181153           0.000000        0.0        0.000000        0.000000   

            CC_DPD_120  CC_BALANCE_150  CC_PAYMENT_150  CC_DPD_150  \
SK_ID_CURR                                                           
394447             0.0       -0.415697       -0.215746         0.0   
361282             0.0       -1.074370        0.280581         0.0   
131335             0.0       -0.817702       -0.514661         0.0   
436351             0.0        0.000000        0.000000         0.0   
181153             0.0        0.000000        0.000000         0.0   

            CC_BALANCE_180  CC_PAYMENT_180  CC_DPD_180  AMT_ANNUITY  \
SK_ID_CURR                                                            
394447           -0.455389        1.318045         0.0       2250.0   
361282           -1.169080       -1.152838         0.0       2250.0   
131335           -1.619869        2.030866         0.0       2250.0   
436351            0.000000        0.000000         0.0       2250.0   
181153            0.000000        0.000000         0.0       2250.0   

            AMT_APPLICATION  AMT_CREDIT  AMT_GOODS_PRICE  
SK_ID_CURR                                                
394447              45000.0     45000.0          45000.0  
361282              45000.0     45000.0          45000.0  
131335              45000.0     45000.0          45000.0  
436351              45000.0     45000.0          45000.0  
181153              45000.0     45000.0          45000.0

### Zusammenführung und Speicherung der Daten

In [55]:
pa_cc = result

In [56]:
pa_cc = pd.merge(pa_cc, cats, how="left", left_index=True, right_index=True)

In [57]:
pa_cc = pa_cc.add_prefix("PA_")

In [58]:
pa_cc2 = pd.read_csv(DATASET_DIR / "2. Datenaufbereitung" / "pa_cc.csv",index_col=0)

In [59]:
pa_cc2 = pa_cc[pa_cc2.columns.values]

In [60]:
pa_cc.head()

PA_CNT_CREDITS  PA_CC_BALANCE_30  PA_CC_PAYMENT_30  PA_CC_DPD_30  \
SK_ID_CURR                                                                     
394447                   1         -1.206367         -1.389108           0.0   
361282                   1          0.946004          0.850318           0.0   
131335                   1          0.722068         -0.427910           0.0   
436351                   1          0.000000          0.000000           0.0   
181153                   1          0.000000          0.000000           0.0   

            PA_CC_BALANCE_60  PA_CC_PAYMENT_60  PA_CC_DPD_60  \
SK_ID_CURR                                                     
394447              1.434002          1.008387           0.0   
361282              0.828945          1.109609           0.0   
131335              0.719445         -0.427508           0.0   
436351              0.000000          0.000000           0.0   
181153              0.000000          0.000000           0.0   

            PA_CC_BALANCE_90  PA_CC_PAYMENT_90  PA_CC_DPD_90  \
SK_ID_CURR                                                     
394447              0.992963         -0.429381           0.0   
361282              0.861611          0.179508           0.0   
131335              0.791575         -0.445538           0.0   
436351              0.000000          0.000000           0.0   
181153              0.000000          0.000000           0.0   

            PA_CC_BALANCE_120  PA_CC_PAYMENT_120  PA_CC_DPD_120  \
SK_ID_CURR                                                        
394447              -0.349513          -0.292196            0.0   
361282              -0.393110          -1.267179            0.0   
131335               0.204483          -0.215249            0.0   
436351               0.000000           0.000000            0.0   
181153               0.000000           0.000000            0.0   

            PA_CC_BALANCE_150  PA_CC_PAYMENT_150  PA_CC_DPD_150  \
SK_ID_CURR                                                        
394447              -0.415697          -0.215746            0.0   
361282              -1.074370           0.280581            0.0   
131335              -0.817702          -0.514661            0.0   
436351               0.000000           0.000000            0.0   
181153               0.000000           0.000000            0.0   

            PA_CC_BALANCE_180  PA_CC_PAYMENT_180  PA_CC_DPD_180  \
SK_ID_CURR                                                        
394447              -0.455389           1.318045            0.0   
361282              -1.169080          -1.152838            0.0   
131335              -1.619869           2.030866            0.0   
436351               0.000000           0.000000            0.0   
181153               0.000000           0.000000            0.0   

            PA_AMT_ANNUITY  PA_AMT_APPLICATION  PA_AMT_CREDIT  \
SK_ID_CURR                                                      
394447              2250.0             45000.0        45000.0   
361282              2250.0             45000.0        45000.0   
131335              2250.0             45000.0        45000.0   
436351              2250.0             45000.0        45000.0   
181153              2250.0             45000.0        45000.0   

            PA_AMT_GOODS_PRICE  PA_CNT_NAME_CONTRACT_TYPE_Cash loans  \
SK_ID_CURR                                                             
394447                 45000.0                                     4   
361282                 45000.0                                     0   
131335                 45000.0                                     3   
436351                 45000.0                                     8   
181153                 45000.0                                     0   

            PA_CNT_NAME_CONTRACT_TYPE_Consumer loans  \
SK_ID_CURR                                             
394447                                             1   
361282        

In [61]:
pa_cc.to_csv(DATASET_DIR / "2. Datenaufbereitung" / "pa_cc_all.csv")